File: scala/samples/src/main/scala/Combinations.scala
def combinations[T](xs: Seq[T], n: Int): Seq[Seq[T]] =
  if n < 0 then throw Exception(s"Invalid length: ${n}")

  if xs.length < n then Seq()
  else if n == 0 then Seq(Seq())
  else
    xs match
      case x +: xs =>
        combinations(xs, n - 1).map(x +: _) ++ combinations(xs, n)
      case Seq() => Seq()

@main def runCombinations =
  println(combinations("abcd", 3))

File: python/combinations.py

def combinations(xs, n):
    if n < 0:
        raise Exception(f"Invalid length: {n}")
    
    if len(xs) < n:
        return 

    if n == 0:
        yield []
        return

    for x in combinations(xs[1:], n - 1):
        yield [xs[0]] + x

    for x in combinations(xs[1:], n):
        yield x

